ghcup-0.0.7.yaml

```diff
@@ -36,12 +36,12 @@ toolRequirements:
           - build-essential
           - curl
           - libffi-dev
-          - libffi6
+          - libffi8
```

Comment:
I think we can completely remove all the `libBlah123` packages if we have the corresponding `libBlah-dev` package.

Reply:
That's not how this works. GHC versions depend on certain specific libraries with specific SONAMEs. We're making a best effort here (should probably also specify libffi7 if it still exists).

Reply:
> That's not how this works. GHC versions depend on certain specific libraries with specific SONAMEs. We're making a best effort here (should probably also specify libffi7 if it still exists).

But this dependency would come from the `-dev` package and the debian/ubuntu base image you are going to use to build ghc. In Debian/Ubuntu, for every release we have just _one_ single soname version, having multiple ones might happen (rarely) with wide used libraries (like ncurses), but just for transition purposes.

Reply:
We support GHC versions back to 8.0.2. The older bindists are linked against older libraries that are **not** corresponding to what `-dev` installs. So that is insufficient. That is not a transition period.

Reply:
I re-checked experimentally, 'cause I do not use Debian/Ubuntu stable releases since a lot of years (just Debian _testing_/_unstable_/_experimental_ and Ubuntu devel releases), with the following command:
```shell
$ bash -c 'for image in debian:10 debian:11 debian:12 debian:testing ubuntu:18.04 ubuntu:20.04 ubuntu:22.04 ubuntu:23.10 ; do echo "Testing $image" ; docker run --quiet --rm "$image" sh -c "apt -qy update ; apt -qy upgrade ; apt -qy install build-essential curl libffi-dev libgmp-dev libncurses-dev libnuma-dev ; curl -sSf https://get-ghcup.haskell.org | env BOOTSTRAP_HASKELL_NONINTERACTIVE=1 BOOTSTRAP_HASKELL_MINIMAL=1 sh ; . /root/.ghcup/env ; ghcup upgrade ; ghcup install ghc 9.4.8" ; done'
```
And it works with all the base images.

Reply:
Because you're installing only one recent GHC version, not historical ones.

Reply:
> We support GHC versions back to 8.0.2. The older bindists are linked against older libraries that are **not** corresponding to what `-dev` installs. So that is insufficient. That is not a transition period.

It means that the dev environment used to build that ghc release was broken or badly configured.
I think it would be more helpful to just deprecate those particular tarballs or rebuild them in the correct way and remove all the corner cases, they just create confusion. I understand that this would be probably a not so easy job to do. Anyway, thanks for fixing the main problem.

Reply:
> It means that the dev environment used to build that ghc release was broken or badly configured.

Not at all. You can't build GHC against libraries that are not released yet. How does that work?

> I think it would be more helpful to just deprecate those particular tarballs or rebuild them in the correct way and remove all the corner cases, they just create confusion. I understand that this would be probably a not so easy job to do. Anyway, thanks for fixing the main problem

I will not do that, unless I get employed by the Haskell Foundation. You can ask GHC devs. A ticket is already open here: https://github.com/haskell/ghcup-hs/issues/903

Reply:
> > It means that the dev environment used to build that ghc release was broken or badly configured.
> 
> Not at all. You can't build GHC against libraries that are not released yet. How does that work?

I expressed myself in a bad way.
I was just thinking that for creating a binary distribution of ghc for distros that has the concept of a devel package (like Debian and Fedora that use packages like `library-dev` or `library-devel`), I would those devel packages and the related binary packages.
I would not create corner cases where, for using a particular bindist of ghc for a particular distro release, you would need to depend on a library version that is not the main version in the same distro release.
We might have exceptions (`ncurses` has been an example in Debian), but I would never consider packaging a ghc bindist that depends on _exceptions_, it would make the releasing process more complicated.

Anyway, as you suggested below, this is not the right place to talk about those ideas.

> 
> > I think it would be more helpful to just deprecate those particular tarballs or rebuild them in the correct way and remove all the corner cases, they just create confusion. I understand that this would be probably a not so easy job to do. Anyway, thanks for fixing the main problem
> 
> I will not do that, unless I get employed by the Haskell Foundation. You can ask GHC devs. A ticket is already open here: [haskell/ghcup-hs#903](https://github.com/haskell/ghcup-hs/issues/903)

I'm sorry, I was not asking you for it :)
Nice to know that we already have a ticket. Thanks for sharing.

Reply:
+1 to removing corner cases, which is why I would recommend against rebuilding old bindists in the first place.

GHC bindists represent snapshots in time, do they not? A particular set of GHC variants build against a particular set of platforms at a particular moment in time. Is that so bad?

Having a different process that continually updates old versions of GHC against new platforms sounds like it would be even better, of course. But to build such a thing would require a lot of resources for development and maintenance that I don't think we have and goes far beyond what I expect of any open source project. Applying that process retroactively to previously-released versions introduces even more ongoing maintenance work. There are  more important actions we can take to improve GHC stability and we have to be very careful how we spend our limited resources.

Reply:
@chreekat indeed, I also wouldn't expect GHC HQ to do that. In fact, I'd rather be completely independent of their CI and bindists. I did draft a HF proposal about this, but I never submitted it, since I wouldn't be able to accept the funding or dedicate enough time for it: https://gist.github.com/hasufell/18cb5438ce7c2ba388160588d751b32d